DataSource wrapping breaks Spring TestContext cache when 3+ context signatures coexist

## Summary

When multiple `@SpringBootTest` classes with different cache signatures run in
the same JVM, the `DataSource` of a cached `ApplicationContext` is closed
prematurely. Tests using that context then fail with
`HikariDataSource (HikariPool-N) has been closed.` even though the
test-context cache should keep it alive.

This happens whenever `query-audit-spring-boot-starter` is on the classpath.
The `@QueryAudit` annotation does **not** need to be present — the
autoconfigured `BeanPostProcessor` alone is enough to trigger it.

## Location

`query-audit-spring-boot-starter`:
- `io.queryaudit.spring.QueryAuditAutoConfiguration` — anonymous inner
  `BeanPostProcessor` (`$1`) that wraps every `DataSource` bean via
  `DataSourceProxyFactory.wrap(ds, interceptor)`.

## Reproduction

Spring Boot 4.0.6, Java 17, JUnit 5, MySQL 8 (real container, not Testcontainers).

Four test classes in the same module:

| Class | Annotations |
|---|---|
| `AppTests` | `@SpringBootTest`, `@ActiveProfiles("test")` |
| `RepoTest` | `@SpringBootTest`, `@ActiveProfiles("test")`, `@Transactional` |
| `ServiceTest` | `@SpringBootTest`, `@ActiveProfiles("test")`, `@Transactional` |
| `ControllerTest` | `@SpringBootTest`, `@AutoConfigureMockMvc`, `@ActiveProfiles("test")`, `@Transactional` |

`build.gradle`:
```
testImplementation 'io.github.haroya01:query-audit-spring-boot-starter:0.3.0'
testImplementation 'io.github.haroya01:query-audit-mysql:0.3.0'
```

Run `./gradlew test`:

```
org.springframework.transaction.CannotCreateTransactionException: Could not open JPA EntityManager for transaction
Caused by: org.hibernate.exception.GenericJDBCException: Unable to acquire JDBC Connection [HikariDataSource (HikariPool-1) has been closed.]
Caused by: java.sql.SQLException: HikariDataSource (HikariPool-1) has been closed.
    at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:95)
    at net.ttddyy.dsproxy.proxy.ProxyLogicSupport.proceedExecution(ProxyLogicSupport.java:72)
```

`@QueryAudit` is **not** required to reproduce.

### Bisecting

| Combination | Result |
|---|---|
| `ServiceTest` alone | pass |
| `RepoTest` + `ServiceTest` | pass |
| `ControllerTest` + `ServiceTest` | pass |
| `AppTests` + `ServiceTest` | pass |
| All four | `ServiceTest` fails |
| All four, with `query-audit` removed | pass |

Failure consistently appears when 3+ distinct context cache signatures coexist.

## Expected vs actual

| | |
|---|---|
| Expected | `DataSource` of a context kept in the test-context cache stays usable as long as the context itself is cached |
| Actual   | underlying `HikariDataSource` closes before the cached context is evicted |

## Impact

Any project depending on `query-audit-spring-boot-starter` with 3+
`@SpringBootTest` classes of distinct cache signatures will see failing
tests, even without using `@QueryAudit`. The autoconfigured
`BeanPostProcessor` makes this opt-out only by removing the dependency.

## Suspected causes (not verified)

- `ProxyDataSource` (net.ttddyy.dsproxy) wrap may close the underlying
  `DataSource` on context shutdown, while the test-context cache still
  holds references in other contexts.
- `io.queryaudit.junit5.QueryAuditDataSourceStore` keeps a static
  `ThreadLocal<QueryInterceptorHolder>`. On same-thread test execution,
  `clear()` from one class may interfere with cached contexts of another.

## Environment

- Spring Boot 4.0.6, Spring Framework 7.x
- query-audit 0.3.0 (`spring-boot-starter`, `mysql`, `junit5`, `core`)
- JUnit 5 (Jupiter)
- Java 17, macOS

--- a/query-audit-spring-boot-starter/build.gradle
+++ b/query-audit-spring-boot-starter/build.gradle
@@ -1,3 +1,19 @@
+// Issue #134 is reported against Spring Boot 4.x while this module compiles and tests
+// against the 3.4.x baseline. The boot4Test source set runs the reported multi-context
+// scenario on an actual Boot 4.x classpath, isolated from the regular test task so the
+// baseline toolchain and the verification toolchain never mix.
+sourceSets {
+    boot4Test {
+        compileClasspath += sourceSets.main.output
+        runtimeClasspath += sourceSets.main.output
+    }
+}
+
+configurations {
+    boot4TestImplementation.extendsFrom implementation
+    boot4TestRuntimeOnly.extendsFrom runtimeOnly
+}
+
 dependencies {
     api project(':query-audit-core')
     api project(':query-audit-junit5')
@@ -14,4 +30,25 @@ dependencies {
     testImplementation 'org.springframework.boot:spring-boot-starter-jdbc:3.4.1'
     testImplementation 'com.zaxxer:HikariCP:5.1.0'
     testImplementation 'com.h2database:h2:2.3.232'
+
+    boot4TestImplementation 'org.springframework.boot:spring-boot-starter-data-jpa:4.0.6'
+    boot4TestImplementation 'org.springframework.boot:spring-boot-starter-web:4.0.6'
+    boot4TestImplementation 'org.springframework.boot:spring-boot-starter-test:4.0.6'
+    // Boot 4.x split MockMvc test support out of spring-boot-test-autoconfigure
+    boot4TestImplementation 'org.springframework.boot:spring-boot-webmvc-test:4.0.6'
+    boot4TestImplementation 'com.h2database:h2:2.3.232'
+    boot4TestRuntimeOnly 'org.junit.platform:junit-platform-launcher'
+}
+
+tasks.register('boot4Test', Test) {
+    description = 'Runs the issue #134 multi-context lifecycle scenario against Spring Boot 4.x.'
+    group = 'verification'
+    testClassesDirs = sourceSets.boot4Test.output.classesDirs
+    classpath = sourceSets.boot4Test.runtimeClasspath
+    useJUnitPlatform()
+    // Cache size 1 forces TestContext eviction between every signature switch — the
+    // cache pressure the #134 report only reached at 49 test classes against the
+    // default cache size of 32. (Pool creation count confirms contexts are rebuilt.)
+    systemProperty 'spring.test.context.cache.maxSize', '1'
+    shouldRunAfter tasks.named('test')
 }

--- a/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/AppTests.java
+++ b/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/AppTests.java
@@ -1,0 +1,21 @@
+package io.queryaudit.spring.boot4;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.ActiveProfiles;
+
+/** Context signature 1 of the issue #134 reproduction: plain {@code @SpringBootTest}. */
+@SpringBootTest
+@ActiveProfiles("test")
+class AppTests {
+
+  @Autowired private ItemService service;
+
+  @Test
+  void contextLoadsAndQueries() {
+    assertThat(service.findAll()).isNotNull();
+  }
+}

--- a/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/Boot4Application.java
+++ b/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/Boot4Application.java
@@ -1,0 +1,12 @@
+package io.queryaudit.spring.boot4;
+
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+/**
+ * Minimal JPA + web application used to reproduce the multi-context DataSource lifecycle scenario
+ * from issue #134 on an actual Spring Boot 4.x classpath. The QueryAudit autoconfiguration is
+ * picked up from the starter's {@code AutoConfiguration.imports} exactly as in a user project — no
+ * {@code @QueryAudit} annotation anywhere, matching the report.
+ */
+@SpringBootApplication
+public class Boot4Application {}

--- a/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/ControllerTest.java
+++ b/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/ControllerTest.java
@@ -1,0 +1,30 @@
+package io.queryaudit.spring.boot4;
+
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * Context signature 2 of the issue #134 reproduction: {@code @AutoConfigureMockMvc} changes the
+ * TestContext cache key, so this class gets its own ApplicationContext and its own connection pool.
+ */
+@SpringBootTest
+@AutoConfigureMockMvc
+@ActiveProfiles("test")
+@Transactional
+class ControllerTest {
+
+  @Autowired private MockMvc mockMvc;
+
+  @Test
+  void itemsEndpointHitsTheDatabase() throws Exception {
+    mockMvc.perform(get("/items")).andExpect(status().isOk());
+  }
+}

--- a/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/Item.java
+++ b/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/Item.java
@@ -1,0 +1,30 @@
+package io.queryaudit.spring.boot4;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+
+@Entity
+public class Item {
+
+  @Id
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  private Long id;
+
+  private String name;
+
+  protected Item() {}
+
+  public Item(String name) {
+    this.name = name;
+  }
+
+  public Long getId() {
+    return id;
+  }
+
+  public String getName() {
+    return name;
+  }
+}

--- a/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/ItemController.java
+++ b/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/ItemController.java
@@ -1,0 +1,20 @@
+package io.queryaudit.spring.boot4;
+
+import java.util.List;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class ItemController {
+
+  private final ItemService service;
+
+  public ItemController(ItemService service) {
+    this.service = service;
+  }
+
+  @GetMapping("/items")
+  public List<Item> items() {
+    return service.findAll();
+  }
+}

--- a/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/ItemRepository.java
+++ b/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/ItemRepository.java
@@ -1,0 +1,5 @@
+package io.queryaudit.spring.boot4;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface ItemRepository extends JpaRepository<Item, Long> {}

--- a/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/ItemService.java
+++ b/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/ItemService.java
@@ -1,0 +1,25 @@
+package io.queryaudit.spring.boot4;
+
+import java.util.List;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+public class ItemService {
+
+  private final ItemRepository repository;
+
+  public ItemService(ItemRepository repository) {
+    this.repository = repository;
+  }
+
+  @Transactional
+  public Item create(String name) {
+    return repository.save(new Item(name));
+  }
+
+  @Transactional(readOnly = true)
+  public List<Item> findAll() {
+    return repository.findAll();
+  }
+}

--- a/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/RepoTest.java
+++ b/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/RepoTest.java
@@ -1,0 +1,27 @@
+package io.queryaudit.spring.boot4;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * Context signature 3 of the issue #134 reproduction: {@code @Transactional} test around the
+ * repository. Runs before {@link ServiceTest} (class-name order) and shares its cached context.
+ */
+@SpringBootTest
+@ActiveProfiles("test")
+@Transactional
+class RepoTest {
+
+  @Autowired private ItemRepository repository;
+
+  @Test
+  void saveAndLoad() {
+    repository.save(new Item("repo"));
+    assertThat(repository.findAll()).isNotEmpty();
+  }
+}

--- a/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/ServiceTest.java
+++ b/query-audit-spring-boot-starter/src/boot4Test/java/io/queryaudit/spring/boot4/ServiceTest.java
@@ -1,0 +1,33 @@
+package io.queryaudit.spring.boot4;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * The reported victim of issue #134: runs last (class-name order), reuses the context cached by
+ * {@link RepoTest}, and in the report failed with {@code HikariDataSource has been closed} because
+ * a sibling context's shutdown had cascaded into this context's pool.
+ */
+@SpringBootTest
+@ActiveProfiles("test")
+@Transactional
+class ServiceTest {
+
+  @Autowired private ItemService service;
+
+  @Test
+  void createAndList() {
+    service.create("svc");
+    assertThat(service.findAll()).isNotEmpty();
+  }
+
+  @Test
+  void listAgainOnReusedContext() {
+    assertThat(service.findAll()).isNotNull();
+  }
+}

--- a/query-audit-spring-boot-starter/src/boot4Test/resources/application-test.properties
+++ b/query-audit-spring-boot-starter/src/boot4Test/resources/application-test.properties
@@ -1,0 +1,8 @@
+# Shared in-memory database across all cached contexts, mirroring the single shared
+# MySQL instance in the issue #134 report. DB_CLOSE_DELAY keeps the schema alive
+# independent of any one pool's connections.
+spring.datasource.url=jdbc:h2:mem:boot4shared;DB_CLOSE_DELAY=-1;MODE=MySQL
+spring.datasource.username=sa
+spring.datasource.password=
+spring.datasource.driver-class-name=org.h2.Driver
+spring.jpa.hibernate.ddl-auto=update

--- a/query-audit-spring-boot-starter/src/boot4Test/resources/junit-platform.properties
+++ b/query-audit-spring-boot-starter/src/boot4Test/resources/junit-platform.properties
@@ -1,0 +1,3 @@
+# Deterministic class order so the scenario always runs as
+# AppTests -> ControllerTest -> RepoTest -> ServiceTest (the reported victim last).
+junit.jupiter.testclass.order.default=org.junit.jupiter.api.ClassOrderer$ClassName

